[IMP] l10n_es: add new 2023 taxes and mod 303 tags

dylankiss opened the pull request:
- Added tags for the new Modelo 303 tax grids: 150, 152, 153, 155, 156, 158
- Added new 0% and 5% taxes (used for ceratain food and electricity)
- Added new 0% and 0.62% Recargo de Equivalencia taxes
- Added fiscal position and Modelo 303 tag mappings for all new taxes

References:
- https://sede.agenciatributaria.gob.es/static_files/Sede/Disenyo_registro/DR_300_399/archivos_23/DR303e23.xlsx
- https://www.boe.es/buscar/doc.php?lang=es&id=BOE-A-2022-22685

Thanks to @pedrobaeza for the input

[FIX] account: support inherited CoA in tax update script

For companies having a Chart of Accounts that inherits from a parent CoA linked to the taxes, the current tax update script did not work.
This change also supports this case.

[task-2993087](https://www.odoo.com/web#id=2993087&cids=1&menu_id=4720&action=333&active_id=967&model=project.task&view_type=form)

Related to: https://github.com/odoo/enterprise/pull/32866

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/103361).

pedrobaeza commented:
Thanks for the info. I'll check it and tell you back.

MiquelRForgeFlow commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -4162,4 +4187,74 @@
     </record>
+
+    <record id="account_tax_template_s_iva0b_150" model="account.tax.template">
+        <field name="description"/> <!-- for resetting the value on existing DBs -->
```
this is not needed, as it's a new tax 😉

dylankiss commented:
> Thanks for the info. I'll check it and tell you back.

@pedrobaeza I found this on the Spanish Agencia Tributatia website on the 0% rate: https://sede.agenciatributaria.gob.es/Sede/en_gb/ayuda/manuales-videos-folletos/manuales-practicos/manual-iva-2022/capitulo-04-sujetos-pasivos-repercusion-impositivo/tipo-impositivo/tipo-impositivo-0-ciento.html

It seems the 0% rate is only used for COVID related expenses. So maybe I should rename it?

pedrobaeza commented:
@acysos the inclusion in model 303 of the 0% taxes field were mentioned in the AEAT meeting? What are the taxes we should include there? I think no new taxes should be put. They would be exempt or not subjected.

pedrobaeza commented:
@dylankiss it seems we finally know the target for this, although it doesn't seem the same. Please check my PR #108868 with all the needed changes. The tags are still missing, waiting for the final call on the AEAT.

dylankiss commented:
@pedrobaeza I just returned from holidays, thanks for your patience. I removed the 0% taxes in my PR, so we can use your PR afterwards. Also updated the Modelo 303 (https://github.com/odoo/enterprise/pull/32866) and trying to get it merged asap.

pedrobaeza commented:
OK, thanks and welcome back. Let's see if the AEAT finally set the definitive format for January.

oco-odoo commented on addons/l10n_es/data/account_tax_data.xml:
```diff
@@ -2080,12 +2110,12 @@
                 'factor_percent': 100,
                 'repartition_type': 'tax',
-                'tag_ids': [ref('mod_303_03')],
+                'tag_ids': [ref('mod_303_155')],
```
You'll need to update the taxes configuration with a local migration script, and change the module version number, just like Claire did with LU here https://github.com/odoo/odoo/pull/108571/files (brand new feature, but let's use it ! :) )

dylankiss replied:
@oco-odoo I added the migration script using Claire's magic function :magic_wand:

dylankiss replied:
@oco-odoo Migration script has been added and tested. Meanwhile fixed a missed case in the migration script as well.

oco-odoo replied:
Nice catch, thanks :)

dylankiss commented:
@jco-odoo

jco-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -160,12 +160,12 @@ def _notify_accountant_managers(taxes_to_check):
     env = api.Environment(cr, SUPERUSER_ID, {})
     chart_template_id = env['ir.model.data'].xmlid_to_res_id(chart_template_xmlid)
-    companies = env['res.company'].search([('chart_template_id', '=', chart_template_id)])
+    companies = env['res.company'].search(['|', ('chart_template_id', '=', chart_template_id), ('chart_template_id.parent_id', '=', chart_template_id)])
```
@dylankiss I am confused here.  Why do we pass common as the template and then suppose only one child level?  Maybe we could use the child_of param here (or just in a for to calculate all the children of the chart_template_id) and if we pass common, we only update the tax templates of the common template.  (so we could just call with another subtemplate if we change something there)

dylankiss replied:
@jco-odoo You're right on both remarks. I changed it to use the `child_of` operator now and only search for tax templates in the given chart.

oco-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -160,12 +160,12 @@ def _notify_accountant_managers(taxes_to_check):
     for company in companies:
         template_to_tax = _get_template_to_real_xmlid_mapping(company, 'account.tax')
-        templates = env['account.tax.template'].with_context(active_test=False).search([("chart_template_id", "=", chart_template_id)])
+        templates = env['account.tax.template'].with_context(active_test=False).search([('chart_template_id', '=', chart_template_id)])
```
useless diff

dylankiss replied:
Removed

oco-odoo commented on addons/account/models/chart_template.py:
```diff
@@ -160,12 +160,12 @@ def _notify_accountant_managers(taxes_to_check):
     env = api.Environment(cr, SUPERUSER_ID, {})
     chart_template_id = env['ir.model.data'].xmlid_to_res_id(chart_template_xmlid)
-    companies = env['res.company'].search([('chart_template_id', '=', chart_template_id)])
+    companies = env['res.company'].search(['|', ('chart_template_id', '=', chart_template_id), ('chart_template_id', 'child_of', chart_template_id)])
```
Wasn't it originally in a distinct commit ? It should

jco-odoo commented:
@robodoo r+ rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

jco-odoo commented:
@dylankiss I think the child_of includes =, but it is a detail, let's see if we can still do the fw-port today.

jco-odoo commented:
@robodoo r+

jco-odoo commented:
@robodoo r+

pedrobaeza commented on addons/l10n_es/data/account_fiscal_position_template_data.xml:
```diff
@@ -521,6 +653,30 @@
+        <record id="fptt_recargo_buy_0b"
+            model="account.fiscal.position.tax.template">
+            <field name="position_id" ref="fp_recargo"/>
+            <field name="tax_src_id" ref="account_tax_template_p_iva0_bc"/>
```
@dylankiss I have detected that these lines should use `account_tax_template_p_iva0_s_bc` instead of `account_tax_template_p_iva0_bc` ("soportado 0%" instead of "exento").
